### ISSUE TITLE
Unicorn 2 regression: ARM pc after prefetch abort is now +4

### ISSUE THREAD

**Issue by gerph** [unicorn-engine/unicorn#1536]
My OS is largely working correctly, but tests on it highlighted a difference in behaviour for the pre-fetch abort behaviour. This appears to be caused by a difference in the Unicorn 1/2 PC value after a UC_ERR_FETCH_UNMAPPED.

* In Unicorn 1, the pc after an unmapped execution was the address that failed to fetch.
* In Unicorn 2, the pc is the address + 4.

## Source environment

I built unicorn on the dev branch at sha 6ed221439916f7d7019047afd7552434308ef658.
This is the most recent version this afternoon, after the recent fix for hflags.
This has been tested on OSX system on Intel hardware (10.14.6)

## Test code

The following test code exercses the 3 cases of aborts that I think are relevant here:

* data abort (a memory access failed)
* undefined instruction abort (a bad instruction was executed)
* pre-fetch abort (a memory access failed to fetch an instruction)

In each case the test code sets up hooks and reports the address from the hook (if given) and the PC at the time of the hook's call. And then on return from `emu_start` we report the error number returned, and the state of the registers.

The intention of the test code is to be able to compare the behaviour between the different hooks and return state, and then between the different versions of Unicorn.

```python
#!/usr/bin/env python
# Sample code for ARM of Unicorn. Nguyen Anh Quynh <aquynh@gmail.com>
# Python sample ported by Loi Anh Tuan <loianhtuan@gmail.com>

import sys
import struct

import capstone

import unicorn
from unicorn import (
        uc_version,
        UcError,
        Uc,
        UC_ARCH_ARM, UC_MODE_ARM,
        UC_ERR_EXCEPTION,
        UC_VERSION_MAJOR, UC_VERSION_MINOR, UC_VERSION_EXTRA,
        UC_HOOK_MEM_READ_UNMAPPED,
        UC_HOOK_MEM_WRITE_UNMAPPED,
        UC_HOOK_MEM_FETCH_UNMAPPED,
        UC_HOOK_INSN_INVALID)
from unicorn.arm_const import (UC_ARM_REG_R0,
        UC_ARM_REG_R1,
        UC_ARM_REG_R2,
        UC_ARM_REG_R3,
        UC_ARM_REG_R4,
        UC_ARM_REG_R5,
        UC_ARM_REG_R6,
        UC_ARM_REG_R7,
        UC_ARM_REG_R8,
        UC_ARM_REG_R9,
        UC_ARM_REG_R10,
        UC_ARM_REG_R11,
        UC_ARM_REG_R12,
        UC_ARM_REG_R13,
        UC_ARM_REG_R14,
        UC_ARM_REG_R15,
        UC_ARM_REG_SP,
        UC_ARM_REG_LR,
        UC_ARM_REG_PC,
        UC_ARM_REG_CPSR,
        UC_ARM_REG_SPSR)

reg_map = [
        UC_ARM_REG_R0,
        UC_ARM_REG_R1,
        UC_ARM_REG_R2,
        UC_ARM_REG_R3,
        UC_ARM_REG_R4,
        UC_ARM_REG_R5,
        UC_ARM_REG_R6,
        UC_ARM_REG_R7,
        UC_ARM_REG_R8,
        UC_ARM_REG_R9,
        UC_ARM_REG_R10,
        UC_ARM_REG_R11,
        UC_ARM_REG_R12,
        UC_ARM_REG_SP,
        UC_ARM_REG_LR,
        UC_ARM_REG_PC,
    ]
arm_names = [
        'r0', 'r1', 'r2', 'r3',
        'r4', 'r5', 'r6', 'r7',
        'r8', 'r9', 'r10', 'r11',
        'r12', 'sp', 'lr', 'pc'
    ]
err_names = {getattr(unicorn, name): name for name in dir(unicorn) if name.startswith('UC_ERR_')}


BAD_ADDRESS = 0x9000 # Somewhere there isn't any code
ADDRESS = 0x1000     # Somewhere we'll put some code
ARM_INSTS = """
E5900000 : LDR     r0, [r0]
F7F0A000 : Undefined instruction
"""
ARM_WORDS = []
for line in ARM_INSTS.splitlines():
    if ' ' not in line or line.startswith('#'):
        continue
    word = line.split(' ')[0]
    value = int(word, 16)
    ARM_WORDS.append(struct.pack('<L', value))
ARM_CODE = b''.join(ARM_WORDS)

# Keep a copy of the code we're running.
with open('code', 'wb') as fh:
    fh.write(ARM_CODE)

def dump_registers(uc):
    print("Registers: ")
    for rn in range(0, 16):
        value = uc.reg_read(reg_map[rn])
        sys.stdout.write("  %3s : 0x%08x" % (arm_names[rn], value))
        if rn % 4 == 3:
            sys.stdout.write("\n")
    print("  CPSR = 0x{:08x}".format(uc.reg_read(UC_ARM_REG_CPSR)))
    print("  SPSR = 0x{:08x}".format(uc.reg_read(UC_ARM_REG_SPSR)))


def read_word(uc, address):
    data = uc.mem_read(address, 4)
    word = data[0] | (data[1]<<8) | (data[2]<<16) | (data[3]<<24)
    return word


def diss(codeword, address):
    for i in md.disasm(codeword, address):
        print("0x%08x: %-8s %s" %(i.address, i.mnemonic, i.op_str))


def hook_code(uc, address, size, user_data):
    codeword = uc.mem_read(address, 4)
    diss(codeword, address)


def hook_unmapped(uc, access, address, size, value, data):
    print("HOOK: Access (%i) to unmapped memory at 0x%08x of size %i, with value %i" % (access, address, size, value))
    print("  PC: 0x%08x" % (uc.reg_read(UC_ARM_REG_PC),))


def hook_inst_invalid(uc, data):
    print("HOOK: Invalid instruction")
    print("  PC: 0x%08x" % (uc.reg_read(UC_ARM_REG_PC),))


md = capstone.Cs(capstone.CS_ARCH_ARM, capstone.CS_MODE_ARM)
md.detail = True
md.syntax = capstone.CS_OPT_SYNTAX_NOREGNAME


# Test ARM
def test_arm():
    print("Testing under Unicorn : {!r}".format(uc_version()))
    print("Header version: {!r}".format((UC_VERSION_MAJOR, UC_VERSION_MINOR, UC_VERSION_EXTRA)))

    print("Testing ARM aborts")
    try:
        # Initialize emulator in ARM mode
        mu = Uc(UC_ARCH_ARM, UC_MODE_ARM)

        # initialize machine registers in different modes
        mu.reg_write(UC_ARM_REG_R0, 0x990000)       # Invalid address to read

        mu.mem_map(ADDRESS, 4096)                   # 4K at the execution address
        mu.mem_write(ADDRESS, ARM_CODE)

        mu.hook_add(UC_HOOK_MEM_READ_UNMAPPED |
                    UC_HOOK_MEM_WRITE_UNMAPPED |
                    UC_HOOK_MEM_FETCH_UNMAPPED, hook_unmapped)
        mu.hook_add(UC_HOOK_INSN_INVALID, hook_inst_invalid)

        print("\n--- Executing code at 0x{:08x} (should trigger data abort)".format(ADDRESS))
        try:
            mu.emu_start(ADDRESS, ADDRESS + len(ARM_CODE))
        except UcError as e:
            print("Unicorn exception was raised, number %i (%s)" % (e.errno, err_names.get(e.errno, 'unknown')))
        dump_registers(mu)


        print("\n--- Executing code at 0x{:08x} (should trigger undefined instruction)".format(ADDRESS + 4))
        try:
            mu.emu_start(ADDRESS + 4, BAD_ADDRESS + len(ARM_CODE))
        except UcError as e:
            print("Unicorn exception was raised, number %i (%s)" % (e.errno, err_names.get(e.errno, 'unknown')))
        dump_registers(mu)


        print("\n--- Executing code at 0x{:08x} (should trigger pre-fetch abort)".format(BAD_ADDRESS))
        try:
            mu.emu_start(BAD_ADDRESS, BAD_ADDRESS + len(ARM_CODE))
        except UcError as e:
            print("Unicorn exception was raised, number %i (%s)" % (e.errno, err_names.get(e.errno, 'unknown')))
        dump_registers(mu)

    except UcError as e:
        print("ERROR: %s" % e)
        raise


if __name__ == '__main__':
    test_arm()
```

## Unicorn 1 test output

```
charles@Laputa ~/projects/RO/pyromaniac (master↑2)> lpython test_aborts.py
Testing under Unicorn : (1, 0, 256L)
Header version: (1, 0, 2)
Testing ARM aborts

--- Executing code at 0x00001000 (should trigger data abort)
HOOK: Access (19) to unmapped memory at 0x00990000 of size 4, with value 0
  PC: 0x00001000
Unicorn exception was raised, number 6 (UC_ERR_READ_UNMAPPED)
Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001000
  CPSR = 0x400001d3
  SPSR = 0x00000000

--- Executing code at 0x00009000 (should trigger undefined instruction)
HOOK: Invalid instruction
  PC: 0x00001004
Unicorn exception was raised, number 10 (UC_ERR_INSN_INVALID)
Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001004
  CPSR = 0x400001d3
  SPSR = 0x00000000

--- Executing code at 0x00009000 (should trigger pre-fetch abort)
HOOK: Access (21) to unmapped memory at 0x00009000 of size 1, with value 0
  PC: 0x00009000
Unicorn exception was raised, number 8 (UC_ERR_FETCH_UNMAPPED)
Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00009000
  CPSR = 0x400001d3
  SPSR = 0x00000000
```

Notice the pre-fetch hook size is 1, which isn't correct.

## Unicorn 2 test output

```
(venv) charles@laputa ~/projects/RO/pyromaniac (master↑2)> lpython test_aborts.py
Testing under Unicorn : (2, 0, 33554437L)
Header version: (2, 0, 5)
Testing ARM aborts

--- Executing code at 0x00001000 (should trigger data abort)
HOOK: Access (19) to unmapped memory at 0x00990000 of size 4, with value 0
  PC: 0x00001000
Unicorn exception was raised, number 6 (UC_ERR_READ_UNMAPPED)
Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001000
  CPSR = 0x400001d3
  SPSR = 0x00000000

--- Executing code at 0x00001004 (should trigger undefined instruction)
HOOK: Invalid instruction
  PC: 0x00001004
Unicorn exception was raised, number 10 (UC_ERR_INSN_INVALID)
Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001004
  CPSR = 0x400001d3
  SPSR = 0x00000000

--- Executing code at 0x00009000 (should trigger pre-fetch abort)
HOOK: Access (21) to unmapped memory at 0x00009000 of size 4, with value 0
  PC: 0x00009000
Unicorn exception was raised, number 8 (UC_ERR_FETCH_UNMAPPED)
Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00009004
  CPSR = 0x400001d3
  SPSR = 0x00000000
```

Notice the PC after execution is 0x9004 (which differs from Unicorn 1), and the size is now 4 (which differs from Unicorn 1, but is correct).

## Unicorn 1/2 side-by-side diff

```
charles@Laputa ~/projects/RO/pyromaniac (master↑2)> diff -W160 -y uni1.txt uni2.txt | cat
Testing under Unicorn : (1, 0, 256L)                          | Testing under Unicorn : (2, 0, 33554437L)
Header version: (1, 0, 2)                             | Header version: (2, 0, 5)
Testing ARM aborts                              Testing ARM aborts

--- Executing code at 0x00001000 (should trigger data abort)            --- Executing code at 0x00001000 (should trigger data abort)
HOOK: Access (19) to unmapped memory at 0x00990000 of size 4, with value 0  HOOK: Access (19) to unmapped memory at 0x00990000 of size 4, with value 0
  PC: 0x00001000                                  PC: 0x00001000
Unicorn exception was raised, number 6 (UC_ERR_READ_UNMAPPED)           Unicorn exception was raised, number 6 (UC_ERR_READ_UNMAPPED)
Registers:                                  Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000       r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000       r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000       r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001000      r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001000
  CPSR = 0x400001d3                               CPSR = 0x400001d3
  SPSR = 0x00000000                               SPSR = 0x00000000

--- Executing code at 0x00001004 (should trigger undefined instruction)     --- Executing code at 0x00001004 (should trigger undefined instruction)
HOOK: Invalid instruction                           HOOK: Invalid instruction
  PC: 0x00001004                                  PC: 0x00001004
Unicorn exception was raised, number 10 (UC_ERR_INSN_INVALID)           Unicorn exception was raised, number 10 (UC_ERR_INSN_INVALID)
Registers:                                  Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000       r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000       r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000       r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001004      r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00001004
  CPSR = 0x400001d3                               CPSR = 0x400001d3
  SPSR = 0x00000000                               SPSR = 0x00000000

--- Executing code at 0x00009000 (should trigger pre-fetch abort)       --- Executing code at 0x00009000 (should trigger pre-fetch abort)
HOOK: Access (21) to unmapped memory at 0x00009000 of size 1, with value 0    | HOOK: Access (21) to unmapped memory at 0x00009000 of size 4, with value 0
  PC: 0x00009000                                  PC: 0x00009000
Unicorn exception was raised, number 8 (UC_ERR_FETCH_UNMAPPED)          Unicorn exception was raised, number 8 (UC_ERR_FETCH_UNMAPPED)
Registers:                                  Registers: 
   r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000       r0 : 0x00990000   r1 : 0x00000000   r2 : 0x00000000   r3 : 0x00000000
   r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000       r4 : 0x00000000   r5 : 0x00000000   r6 : 0x00000000   r7 : 0x00000000
   r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000       r8 : 0x00000000   r9 : 0x00000000  r10 : 0x00000000  r11 : 0x00000000
  r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00009000      |   r12 : 0x00000000   sp : 0x00000000   lr : 0x00000000   pc : 0x00009004
  CPSR = 0x400001d3                               CPSR = 0x400001d3
  SPSR = 0x00000000                               SPSR = 0x00000000

```

## Unicorn 1 behaviour

Unicorn 1 has the following behaviour:

* data abort hook reports the address as 0x99000, size 4, at pc = 0x1000. exits with pc = 0x1000. (ie consistent pc)
* undefined instruction hook reports pc = 0x1004. exits with pc = 0x1004. (ie consistent pc)
* prefetch abort hook reports the address as 0x9000, size 1, at pc 0x9000. exits with pc = 0x9000. (ie consistent pc, but a wrong size - the instruction is of size 4 in ARM mode)

## Unicorn 2 behaviour

Unicorn 2 has the following behaviour:

* data abort hook reports the address as 0x99000, size 4, at pc = 0x1000. exits with pc = 0x1000. (ie consistent pc)
* undefined instruction hook reports pc = 0x1004. exits with pc = 0x1004. (ie consistent pc)
* prefetch abort hook reports the address as 0x9000, size 4, at pc 0x9000. exits with pc = 0x9004. (ie *inconsistent* pc on exit, but the size is right)

## Expected behaviour

If we were following the ARM documentation:

* for data abort, the PC after execution would be at +8
* for pre-fetch abort, the PC after execution would be at +4
* for undefined instruction, the PC after execution would be at +4

Reference: https://developer.arm.com/documentation/ddi0344/k/programmers-model/exceptions/aborts and https://developer.arm.com/documentation/ddi0344/k/programmers-model/exceptions/undefined-instruction (but there are probably better references)

Under Unicorn 1, the PC on exit from emu_start is pointing at the instruction that failed. This is almost certainly what most users of an emulation system would expect, I believe. In my operating system, before entering the exception vector entry point, I update the registers to apply these offsets, then call the exception vector entry point.

At present under Unicorn 2, the behaviour is inconsistent, with the pre-fetch abort having the PC set to a different address to the one that failed - the address that is consistent with the ARM documentation, but inconsistent to the behaviour of the other abort types.

If the current behaviour is retained, my OS will need to change the PC modifications that it applies depending on the version of Unicorn that is being used. Obviously I'm ok with making that change, but it will affect anyone else who is trying to handle that case in a similar manner.

It would also be good to include the details of the change in the change information between the versions, as other users would need to be aware of that if they were trying to handle pre-fetch aborts.

If the Unicorn 1 behaviour is restored, that gives consistency to the interface.

The difference in the 'size' parameter on the pre-fetch hook is in my opinion preferred, and should be retained (because the instruction size is 4 bytes), but probably that should be documented, just for completeness. It's less likely that anyone would be affected by that one.

**Comment by wtdcode** [comment]
Thanks for your report. Such behavior between Unicorn1 and Unicorn2 should keep the same.

**Comment by wtdcode** [comment]
Fixed in ea9c7425b028f5f54c33e23044c14c31a8a95813
